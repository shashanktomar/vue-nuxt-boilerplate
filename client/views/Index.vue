<template>
  <main>
    <div>
      <app-logo />
      <action-bar />

      <section
        v-if="hasComments"
        :class="$style.container"
      >
        <comment
          v-for="comment in comments"
          :comment="comment"
          :key="comment.id"
        />
      </section>

    </div>
  </main>
</template>

<script>
// @flow

import Vue from 'vue'
import { Store } from 'vuex'
import Component from 'nuxt-class-component'
import { Getter, State } from 'vuex-class'

import ActionBar from '~/components/ActionBar'
import AppLogo from '~/components/AppLogo'
import Comment from '~/components/Comment'

import type { CommentType, StateType } from '~/types'

// @vue/component
@Component({
  'components': {
    ActionBar,
    AppLogo,
    Comment,
  },
})
/**
 * Main page (or index page).
 * Mounted as `/` by default.
 */
export default class Index extends Vue {
  @State('comments')
  /**
   * List of predownloaded comments, bound from Vuex.
   */
  comments: Array<CommentType>

  @Getter('hasComments')
  /**
   * Returns either we have any comments or not.
   */
  hasComments: boolean

  /**
   * Fetches comments from external API from the server side.
   * This method should preload Vuex store.
   *
   * @see https://nuxtjs.org/api/pages-fetch
   * @param context - Nuxt `context` instance.
   * @param context.store - Current Vuex store.
   * @returns List of downloaded comments.
   */
  fetch ({ store }: { store: Store<StateType> }): Promise<Array<CommentType>> {
    // Uncomment the next line to test flow types:
    // console.log(this.comments + store.state.comments.length)
    return store.dispatch('fetchComments')
  }
}
</script>

<style lang="scss" module>
.container {
  display: flex;
  flex-wrap: wrap;
  justify-content: space-evenly;
}
</style>
