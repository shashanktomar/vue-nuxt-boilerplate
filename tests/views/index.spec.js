// @flow

import { mount, createLocalVue } from '@vue/test-utils'
import MockAdapter from 'axios-mock-adapter'

import Index from '~/views/Index'
import { storeFactory, commentFactory } from '../fixtures/vuex'

const localVue = createLocalVue()

describe('unit tests for Index view', () => {
  let store
  let comments

  beforeEach(() => {
    comments = commentFactory.buildList(3)
    store = storeFactory.build({ 'state': { comments } }, { localVue })
  })

  test('should have three comments', () => {
    const wrapper = mount(Index, { store, localVue, 'propsData': { comments } })
    expect(wrapper.findAll('.commentComponent')).toHaveLength(3)
  })

  test('should load new comments on actions', async () => {
    const comment = comments[0]
    const wrapper = mount(Index, { store, localVue, 'propsData': { comments } })

    const mock = new MockAdapter(store.$axios)
    mock.onGet('/comments').reply(200, [comment])

    await wrapper.vm.$store.dispatch('fetchComments', wrapper.vm)
    expect(wrapper.vm.$store.state.comments).toHaveLength(1)
    expect(wrapper.vm.$store.state.comments[0].id).toEqual(comment.id)
    expect(wrapper.vm.$store.state.comments[0].email).toEqual(comment.email)
    expect(wrapper.findAll('.commentComponent')).toHaveLength(1)
  })
})

describe('snapshot test for Index view', () => {
  let store
  let comments

  beforeAll(() => {
    // We need a seed here to be consistent for snapshot testing:
    comments = commentFactory.buildList(3, {}, { 'seed': 1342 })
    store = storeFactory.build({ 'state': { comments } }, { localVue })
  })

  test('should match the snapshot', () => {
    const wrapper = mount(Index, { store, localVue, 'propsData': { comments } })
    expect(wrapper.html()).toMatchSnapshot()
  })
})
