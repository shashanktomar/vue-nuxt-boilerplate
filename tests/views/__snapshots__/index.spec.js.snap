// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test for Index view should match the snapshot 1`] = `
"<main comments=\\"[object Object],[object Object],[object Object]\\"><div><div class=\\"logo\\"><h1><a href=\\"https://github.com/wemake-services/wemake-vue-template\\">
      wemake-vue-template
    </a></h1> <h2><a href=\\"https://wemake-services.gitbook.io/wemake-vue-template/\\">
      documentation
    </a></h2></div> <div class=\\"actions\\"><button class=\\"reload\\">reload comments</button></div> <section class=\\"container\\"><div class=\\"commentComponent commentNegative\\"><p class=\\"author\\">Alfred.Barton@hotmail.com</p> <p class=\\"body\\">Et perspiciatis facilis sed sit labore. Ipsum mollitia sunt occaecati. Expedita ut perspiciatis quod enim autem eum. Unde quibusdam eligendi deleniti consectetur.</p> <div class=\\"rating\\"><button class=\\"ratingUp\\">
      +
    </button> <span class=\\"number\\">-10</span> <button class=\\"ratingDown\\">
      -
    </button></div></div><div class=\\"commentComponent commentNegative\\"><p class=\\"author\\">Alfred.Barton@hotmail.com</p> <p class=\\"body\\">Et perspiciatis facilis sed sit labore. Ipsum mollitia sunt occaecati. Expedita ut perspiciatis quod enim autem eum. Unde quibusdam eligendi deleniti consectetur.</p> <div class=\\"rating\\"><button class=\\"ratingUp\\">
      +
    </button> <span class=\\"number\\">-10</span> <button class=\\"ratingDown\\">
      -
    </button></div></div><div class=\\"commentComponent commentNegative\\"><p class=\\"author\\">Alfred.Barton@hotmail.com</p> <p class=\\"body\\">Et perspiciatis facilis sed sit labore. Ipsum mollitia sunt occaecati. Expedita ut perspiciatis quod enim autem eum. Unde quibusdam eligendi deleniti consectetur.</p> <div class=\\"rating\\"><button class=\\"ratingUp\\">
      +
    </button> <span class=\\"number\\">-10</span> <button class=\\"ratingDown\\">
      -
    </button></div></div></section></div></main>"
`;
